common/src/project.rs

```diff
@@ -7,62 +7,65 @@ use std::error::Error;
+///
+/// Additionaly, while host segments are technically case-insensitive, the filesystem isn't,
+/// so we restrict project names to be lower case. We also restrict the use of profanity,
+/// as well as a list of reserved words.
 #[derive(Clone, Serialize, Debug, Eq, PartialEq)]
+#[cfg_attr(feature = "backend", derive(sqlx::Type, Hash))]
+#[cfg_attr(feature = "backend", sqlx(transparent))]
 pub struct ProjectName(String);
 
-impl<'de> Deserialize<'de> for ProjectName {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
-    where
-        D: Deserializer<'de>,
-    {
-        let s: String = String::deserialize(deserializer)?;
-        s.parse().map_err(DeError::custom)
-    }
-}
-
-impl std::fmt::Display for ProjectName {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        self.0.fmt(f)
-    }
-}
-
 impl ProjectName {
-    pub fn is_valid(hostname: &str) -> bool {
+    /// The rules a valid project name must follow.
+    pub const RULES: &str = "\
+    Project names must:
+    1. only contain lowercase alphanumeric characters or dashes `-`.
+    2. not start or end with a dash.
+    3. not be empty.
+    4. be shorter than 64 characters.
+    5. not contain any profanities.
+    6. not be a reserved word.\
+    ";
+
+    pub fn is_valid(label: &str) -> bool {
         fn is_valid_char(byte: u8) -> bool {
-            matches!(byte, b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'-' | b'_')
+            matches!(byte, b'a'..=b'z' | b'0'..=b'9' | b'-')
         }
 
-        fn is_profanity_free(hostname: &str) -> bool {
-            let (_censored, analysis) = Censor::from_str(hostname).censor_and_analyze();
+        fn is_profanity_free(label: &str) -> bool {
+            let (_censored, analysis) = Censor::from_str(label).censor_and_analyze();
             !analysis.is(Type::MODERATE_OR_HIGHER)
         }
 
-        fn is_reserved(hostname: &str) -> bool {
+        fn is_reserved(label: &str) -> bool {
             static INSTANCE: OnceCell<HashSet<&str>> = OnceCell::new();
-            INSTANCE.get_or_init(|| HashSet::from(["shuttle.rs"]));
+            INSTANCE.get_or_init(|| HashSet::from(["shuttleapp", "shuttle"]));
 
             INSTANCE
                 .get()
                 .expect("Reserved words not set")
-                .contains(hostname)
+                .contains(label)
         }
 
-        let separators = ['-', '_'];
+        // Each label in a hostname can be between 1 and 63 chars.
+        let is_too_long = label.len() > 63;
 
-        !(hostname.bytes().any(|byte| !is_valid_char(byte))
-            || is_reserved(hostname)
-            || !is_profanity_free(hostname)
-            || hostname.ends_with(separators)
-            || hostname.starts_with(separators)
-            || hostname.is_empty())
+        !(label.bytes().any(|byte| !is_valid_char(byte))
+            || is_reserved(label)
+            || !is_profanity_free(label)
+            || is_too_long
```

Comment:
nit: length check can be first to guard against unnecessary iteration.

Reply:
Great idea, this adapts to different shapes of invalid inputs in a pretty neat way.

Reply:
Did you want to change this?
(P.S it can be: empty || too long || (other)... )

Reply:
Yes, I would. Could do so by tomorrow. Don't hold back if you'd like to fix this now already 🙂